File: src/main/java/com/extendedae_plus/network/CPacketInterfaceScaling.java
package com.extendedae_plus.network;

import appeng.api.stacks.GenericStack;
import appeng.menu.implementations.InterfaceMenu;
import com.extendedae_plus.client.render.widgets.button.EAEPActionItems;
import com.extendedae_plus.network.base.CPacketGeneric;
import com.extendedae_plus.network.base.EAEPNetworkPacket;
import com.glodblock.github.extendedae.container.ContainerExInterface;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.server.level.ServerPlayer;
import net.neoforged.neoforge.network.PacketDistributor;

/**
 * C2S：调整 ME 接口配置槽位(标记物品)的数量。
 * 支持按因子倍增或整除，且保持最小值为 1。
 */
@EAEPNetworkPacket("interface_scaling")
public record CPacketInterfaceScaling(int scale, boolean divide) implements CPacketGeneric {
    public static final StreamCodec<FriendlyByteBuf, CPacketInterfaceScaling> STREAM_CODEC = StreamCodec.composite(
            ByteBufCodecs.INT, CPacketInterfaceScaling::scale,
            ByteBufCodecs.BOOL, CPacketInterfaceScaling::divide,
            CPacketInterfaceScaling::new
    );

    public static void send(EAEPActionItems action) {
        int scale = 0;
        boolean divide = false;
        switch (action) {
            case MUL2 -> scale = 2;
            case MUL3 -> scale = 3;
            case MUL5 -> scale = 5;
            case DIV2 -> {
                scale = 2;
                divide = true;
            }
            case DIV3 -> {
                scale = 3;
                divide = true;
            }
            case DIV5 -> {
                scale = 5;
                divide = true;
            }
        }
        if (scale > 0)
            PacketDistributor.sendToServer(new CPacketInterfaceScaling(scale, divide));
    }

    public void handleServer(final ServerPlayer player) {
        // 支持 AE2 原版接口和 ExtendedAE 扩展接口（若存在）
        InterfaceMenu menu = null;
        ContainerExInterface exMenu = null;
        if (player.containerMenu instanceof InterfaceMenu im) {
            menu = im;
        } else if (player.containerMenu instanceof ContainerExInterface cem) {
            exMenu = cem;
        } else {
            return;
        }

        try {
            var logic = (menu != null ? menu.getHost() : exMenu.getHost()).getInterfaceLogic();
            var config = logic.getConfig();
            // 对所有配置槽统一生效
            int size = config.size();
            for (int index = 0; index < size; index++) {
                var stack = config.getStack(index);
                if (stack == null) continue;

                long amount = stack.amount();
                int scale = this.scale;

                long scaledAmount = 0;
                if (this.divide) {
                    if (amount % scale > 0) continue;
                    scaledAmount = amount / scale;
                } else scaledAmount = amount * scale;

                config.setStack(index, new GenericStack(stack.what(), scaledAmount));
            }
        } catch (Throwable ignored) {
        }
    }
}

File: src/main/kotlin/com/extendedae_plus/network/CPacketInterfaceScaling.kt
package com.extendedae_plus.network

import appeng.api.stacks.GenericStack
import appeng.menu.implementations.InterfaceMenu
import com.extendedae_plus.client.render.widgets.button.EAEPActionItems
import com.fish.fishlib.network.FishNetworkPacket
import com.fish.fishlib.network.PacketStreamCodec
import com.fish.fishlib.network.base.CPacketGeneric
import com.fish.fishlib.network.base.PacketGeneric.Companion.sendToServer
import com.fish.fishlib.util.UtilMath
import com.glodblock.github.extendedae.container.ContainerExInterface
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.server.level.ServerPlayer

/**
 * C2S：调整 ME 接口配置槽位(标记物品)的数量。
 * 支持按因子倍增或整除，且保持最小值为 1。
 */
@FishNetworkPacket("interface_scaling")
@JvmRecord
data class CPacketInterfaceScaling(val scale: Int) : CPacketGeneric {
    companion object {
        @PacketStreamCodec
        val streamCodec: StreamCodec<FriendlyByteBuf, CPacketInterfaceScaling> =
            StreamCodec.composite(
                ByteBufCodecs.INT, CPacketInterfaceScaling::scale,
                ::CPacketInterfaceScaling
            )

        @JvmStatic
        fun send(action: EAEPActionItems) {
            val scale = when (action) {
                EAEPActionItems.Mul2 -> 2
                EAEPActionItems.Mul3 -> 3
                EAEPActionItems.Mul5 -> 5
                EAEPActionItems.Div2 -> -2
                EAEPActionItems.Div3 -> -3
                EAEPActionItems.Div5 -> -5
                else -> return
            }
            CPacketInterfaceScaling(scale).sendToServer()
        }
    }

    override fun handleServer(player: ServerPlayer) {
        val logic = when (val menu = player.containerMenu) {
            is InterfaceMenu -> menu.host
            is ContainerExInterface -> menu.host
            else -> return
        }.interfaceLogic

        val inv = logic.config
        for (indexStack in 0..<inv.size()) {
            var stack = inv.getStack(indexStack) ?: continue
            val amount = UtilMath.scale(stack.amount, this.scale.toLong(), true)

            stack = GenericStack(stack.what, amount)
            inv.setStack(indexStack, stack)
        }
    }
}
